modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/LogicalRelImplementor.java

```diff
@@ -377,37 +384,37 @@ public Node<RowT> visit(IgniteMergeJoin rel) {
 
         SqlJoinProjection<RowT> joinProjection = createJoinProjection(rel, rel.getRowType(), leftType.getFieldCount());
 
-        int pairsCnt = rel.analyzeCondition().pairs().size();
+        ImmutableBitSet nullCompAsEqual = nullComparisonStrategyVector(rel, rel.analyzeCondition().leftSet());
 
-        ImmutableBitSet leftKeys = rel.analyzeCondition().leftSet();
+        ImmutableIntList leftKeys = rel.leftCollation().getKeys();
+        ImmutableIntList rightKeys = rel.rightCollation().getKeys();
 
-        List<RexNode> conjunctions = RelOptUtil.conjunctions(rel.getCondition());
+        // Convert conditions to using collation indexes instead of field indexes.
+        List<IntPair> condIndexes = rel.analyzeCondition().pairs().stream()
```

Comment:
I think stream + linear scan of keys may introduced performance penalty. Can we introduce benchmark to measure this? Perhaps, scenario should be like this: single merge join of two tables with single row in each; the size of join pairs is configurable

Reply:
Rewritten with for-loops.